File: android/src/main/kotlin/baitianwei/flutter_vod_upload_plugin/FlutterVodUploadPlugin.kt
package baitianwei.flutter_vod_upload_plugin

import android.app.Activity
import android.app.Application
import androidx.annotation.NonNull
import com.alibaba.sdk.android.vod.upload.VODUploadCallback
import com.alibaba.sdk.android.vod.upload.VODUploadClientImpl
import com.alibaba.sdk.android.vod.upload.model.UploadFileInfo
import com.alibaba.sdk.android.vod.upload.model.VodInfo
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar


/** FlutterVodUploadPlugin */
public class FlutterVodUploadPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    var uploader: VODUploadClientImpl? = null


    private var channel: MethodChannel? = null
    private var pluginBinding: FlutterPlugin.FlutterPluginBinding? = null
    private var activityBinding: ActivityPluginBinding? = null
    private var application: Application? = null
    private var activity: Activity? = null

    fun setup(messenger: BinaryMessenger, application: Application, activity: Activity) {
        this.activity = activity
        this.application = application
        channel = MethodChannel(messenger, "flutter_vod_upload_plugin")
        channel!!.setMethodCallHandler(this)
        uploader = VODUploadClientImpl(application)
    }

    private fun tearDown() {
        activityBinding = null
        channel?.setMethodCallHandler(null)
        channel = null
        application = null
    }


    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        pluginBinding = flutterPluginBinding
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        pluginBinding = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        this.activityBinding = binding
        setup(
                pluginBinding!!.binaryMessenger,
                (pluginBinding!!.applicationContext as Application),
                activityBinding!!.activity)

    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }


    override fun onDetachedFromActivity() {
        tearDown()
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity()
    }


    companion object {
        @JvmStatic
        fun registerWith(registrar: Registrar) {
            if (registrar.activity() == null) return
            val activity = registrar.activity()
            var application: Application? = null
            if (registrar.context() != null) {
                application = registrar.context().applicationContext as Application
            }
            val plugin = FlutterVodUploadPlugin()
            plugin.setup(registrar.messenger(), application!!, activity)
        }
    }

    private fun invokeMethod(method: String, arguments: Any?) {
        activity?.runOnUiThread {
            channel?.invokeMethod(method, arguments)
        }
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "start" -> start()
            "pause" -> pause()
            "stop" -> stop()
            "resume" -> resume()
            "setListener" -> {
                setUploadAuthAndAddress(call.argument<String>("uploadAuth")!!, call.argument<String>("uploadAddress")!!)
            }
            "addFile" -> {
                val vodInfo = VodInfo()
                val info = call.argument<Map<*, *>>("vodInfo")
                vodInfo.title = info?.get("title") as? String
                if (info?.get("isShowWaterMark") is Boolean) {
                    vodInfo.isShowWaterMark = info["isShowWaterMark"] as? Boolean
                }
                addFile(call.argument<String>("filePath")!!, vodInfo)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun setUploadAuthAndAddress(uploadAuth: String, uploadAddress: String) {

        val callback: VODUploadCallback = object : VODUploadCallback() {
            override fun onUploadSucceed(info: UploadFileInfo?) {
                invokeMethod("onUploadSucceed", null)
            }

            override fun onUploadFailed(info: UploadFileInfo?, code: String?, message: String?) {
                invokeMethod("onUploadFailed", mapOf("code" to code, "message" to message))
            }

            override fun onUploadProgress(info: UploadFileInfo?, uploadedSize: Long, totalSize: Long) {
                invokeMethod("onUploadProgress", mapOf("uploadedSize" to uploadedSize, "totalSize" to totalSize))
            }

            override fun onUploadTokenExpired() {
                invokeMethod("onUploadTokenExpired", null)
            }

            override fun onUploadRetry(code: String?, message: String?) {
                invokeMethod("onUploadRetry", null)
            }

            override fun onUploadRetryResume() {
                invokeMethod("onUploadRetryResume", null)
            }

            override fun onUploadStarted(uploadFileInfo: UploadFileInfo?) {
                invokeMethod("onUploadStarted", null)
                uploader?.setUploadAuthAndAddress(uploadFileInfo, uploadAuth, uploadAddress)
            }
        }
        uploader?.init(callback)
    }

    private fun addFile(filePath: String, vodInfo: VodInfo) {
        uploader?.addFile(filePath, vodInfo)
    }

    private fun start() {
        uploader?.start()
    }

    private fun stop() {
        uploader?.stop()
    }

    private fun pause() {
        uploader?.pause()
    }

    private fun resume() {
        uploader?.resume()
    }
}

File: ios/Classes/FlutterVodUploadPlugin.m
#import "FlutterVodUploadPlugin.h"
#import <VODUpload/VODUploadClient.h>

@implementation FlutterVodUploadPlugin {
    VODUploadClient* uploader;
}
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"flutter_vod_upload_plugin"
                                     binaryMessenger:[registrar messenger]];
    FlutterVodUploadPlugin* instance = [[FlutterVodUploadPlugin alloc] init];
    instance.channel = channel;
    [registrar addMethodCallDelegate:instance channel:channel];
}
- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"setUploadAuthAndAddress" isEqualToString:call.method]) {
        //        [self setUploadAuthAndAddress:<#(UploadFileInfo *)#> uploadAuth:<#(NSString *)#> uploadAddress:<#(NSString *)#>]
    } else if ([@"start" isEqualToString:call.method]) {
        [self start];
    } else if ([@"stop" isEqualToString:call.method]){
        [self pause];
    } else if ([@"pause" isEqualToString:call.method]){
        [self stop];
    } else if ([@"resume" isEqualToString:call.method]){
        [self resume];
    } else if ([@"setListener" isEqualToString:call.method]){
        [self setListener:call.arguments[@"uploadAuth"] uploadAddress:call.arguments[@"uploadAddress"]];
    } else if ([@"addFile" isEqualToString:call.method]){
        NSString* filePath = call.arguments[@"filePath"];
        [self addFile:filePath vodInfo:createVodInfo(call.arguments[@"vodInfo"])];
    } else {
        result(FlutterMethodNotImplemented);
    }
    
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        uploader =[VODUploadClient new];
    }
    return self;
}


static VodInfo* createVodInfo(NSDictionary *dict) {
    VodInfo* vodInfo = [VodInfo new];
    
    vodInfo.title = dict[@"title"];
    if (dict[@"desc"] && dict[@"desc"] != [NSNull null]) {
        vodInfo.desc = dict[@"desc"];
    }
    if (dict[@"cateId"] && dict[@"cateId"] != [NSNull null]) {
        vodInfo.cateId = dict[@"cateId"];
    }
    if (dict[@"tags"] && dict[@"tags"] != [NSNull null]) {
        vodInfo.tags = dict[@"tags"];
    }
    if (dict[@"userData"] && dict[@"userData"] != [NSNull null]) {
        vodInfo.userData = dict[@"userData"];
    }
    if (dict[@"coverUrl"] && dict[@"coverUrl"] != [NSNull null]) {
        vodInfo.coverUrl = dict[@"coverUrl"];
    }
    if (dict[@"isProcess"] && dict[@"isProcess"] != [NSNull null]) {
        vodInfo.isProcess = dict[@"isProcess"];
    }
    
    if (dict[@"isShowWaterMark"] && dict[@"isShowWaterMark"] != [NSNull null]) {
        vodInfo.isShowWaterMark = dict[@"isShowWaterMark"];
    }
    if (dict[@"priority"] && dict[@"priority"] != [NSNull null]) {
        vodInfo.priority = dict[@"priority"];
    }
    
    return vodInfo;
}

static NSDictionary* vodInfoToMap(VodInfo* vodInfo) {
    NSMutableDictionary* dict = @{}.mutableCopy;
    
    if (vodInfo.title) {
        dict[@"title"] = vodInfo.title;
    }
    if (vodInfo.desc) {
        dict[@"desc"] = vodInfo.desc;
    }
    if (vodInfo.cateId) {
        dict[@"cateId"] = vodInfo.cateId;
    }
    if (vodInfo.tags) {
        dict[@"tags"] = vodInfo.tags;
    }
    if (vodInfo.userData) {
        dict[@"userData"] = vodInfo.userData;
    }
    if (vodInfo.coverUrl) {
        dict[@"coverUrl"] = vodInfo.coverUrl;
    }
    if (vodInfo.isProcess) {
        dict[@"isProcess"] = [NSNumber numberWithBool:vodInfo.isProcess] ;
    }
    
    if (vodInfo.isShowWaterMark) {
        dict[@"isShowWaterMark"] = [NSNumber numberWithBool:vodInfo.isShowWaterMark] ;
    }
    if (vodInfo.priority) {
        dict[@"priority"] = vodInfo.priority;
    }
    
    
    return dict;
    
}

static NSDictionary* uploadFileInfoToMap(UploadFileInfo* fileInfo) {
    NSMutableDictionary* dict = @{}.mutableCopy;
    if (fileInfo.filePath) {
        dict[@"filePath"] = fileInfo.filePath;
    }
    if (fileInfo.endpoint) {
        dict[@"endpoint"] = fileInfo.endpoint;
    }
    if (fileInfo.bucket) {
        dict[@"bucket"] = fileInfo.bucket;
    }
    if (fileInfo.object) {
        dict[@"object"] = fileInfo.object;
    }
    if (fileInfo.vodInfo) {
        dict[@"vodInfo"] = vodInfoToMap(fileInfo.vodInfo);
    }
    if (fileInfo.state) {
        switch (fileInfo.state) {
            case VODUploadFileStatusSuccess:
                dict[@"state"] = @"SUCCESS";
                break;
            case VODUploadFileStatusUploading:
                dict[@"state"] = @"Uploading";
                break;
            case VODUploadFileStatusCanceled:
                dict[@"state"] = @"Canceled";
                break;
            case VODUploadFileStatusFailure:
                dict[@"state"] = @"Failure";
            case VODUploadFileStatusPaused:
                dict[@"state"] = @"Paused";
                break;
            default:
                break;
        }
    }
    return dict;
    
}

// MARK: 上传地址和凭证方式

- (void)setListener:(NSString* )uploadAuth
      uploadAddress:(NSString* )uploadAddress {
    
    // weakself
    __weak typeof(self) weakSelf = self;
    // setup callback
    OnUploadFinishedListener onUploadSucceed = ^(UploadFileInfo* fileInfo, VodUploadResult* result){
        [weakSelf.channel invokeMethod:@"onUploadSucceed" arguments:@{
            @"info":uploadFileInfoToMap(fileInfo)
        }];
    };
    OnUploadFailedListener onUploadFailed = ^(UploadFileInfo* fileInfo, NSString *code, NSString* message){
        NSLog(@"upload failed callback code = %@, error message = %@", code, message);
        
        [weakSelf.channel invokeMethod:@"onUploadFailed" arguments:[NSDictionary dictionaryWithObjectsAndKeys:
                                                                    uploadFileInfoToMap(fileInfo),@"info",
                                                                    code,@"code",
                                                                    message, @"message",
                                                                    nil]];
        
    };
    OnUploadProgressListener onUploadProgress = ^(UploadFileInfo* fileInfo, long uploadedSize, long totalSize) {
        
        [weakSelf.channel invokeMethod:@"onUploadProgress" arguments:[NSDictionary dictionaryWithObjectsAndKeys:
                                                                      uploadFileInfoToMap(fileInfo),@"info",
                                                                      [NSNumber numberWithLong: uploadedSize],@"uploadedSize",
                                                                      [NSNumber numberWithLong: totalSize], @"totalSize",
                                                                      nil]];
    };
    OnUploadTokenExpiredListener onUploadTokenExpired = ^{
        NSLog(@"upload token expired callback.");
        [weakSelf.channel invokeMethod:@"onUploadTokenExpired" arguments:nil];
        // token过期，设置新的上传凭证，继续上传
        // [uploader resumeWithAuth:@""]
    };
    OnUploadRertyListener onUploadRetry = ^{
        [weakSelf.channel invokeMethod:@"onUploadRetry" arguments:nil];
    };
    OnUploadRertyResumeListener onUploadRetryResume = ^{
        [weakSelf.channel invokeMethod:@"onUploadRetryResume" arguments:nil];
    };
    OnUploadStartedListener onUploadStarted = ^(UploadFileInfo* fileInfo) {
        [weakSelf.channel invokeMethod:@"onUploadStarted" arguments:@{
            @"info":uploadFileInfoToMap(fileInfo)
        }];
        [self->uploader setUploadAuthAndAddress:fileInfo uploadAuth: uploadAuth uploadAddress:uploadAddress];
    };
    VODUploadListener *listener = [[VODUploadListener alloc] init];
    listener.finish = onUploadSucceed;
    listener.failure = onUploadFailed;
    listener.progress = onUploadProgress;
    listener.expire = onUploadTokenExpired;
    listener.retry = onUploadRetry;
    listener.retryResume = onUploadRetryResume;
    listener.started = onUploadStarted;
    // init with upload address and upload auth
    [uploader setListener:listener];
}

// MARK: 设置上传地址 和 上传凭证
- (BOOL)setUploadAuthAndAddress:(UploadFileInfo *)uploadFileInfo
                     uploadAuth:(NSString *)uploadAuth
                  uploadAddress:(NSString *)uploadAddress {
    return [uploader setUploadAuthAndAddress:uploadFileInfo uploadAuth:uploadAuth uploadAddress:uploadAddress];
}
// MARK: 添加文件到上传列表
- (BOOL)addFile:(NSString *)filePath
        vodInfo:(VodInfo *)vodInfo{
    return [uploader addFile:filePath vodInfo:vodInfo];
}
// MARK: 上传控制
- (BOOL) start {
    return [uploader start];
}
- (BOOL) stop {
    return [uploader stop];
}
- (BOOL) pause {
    return [uploader pause];
}
- (BOOL) resume {
    return [uploader resume];
}
@end
